enhancements/kubelet/master-bootstrap-credentials.md

```diff
@@ -0,0 +1,120 @@
+    called `configmap/kubelet-bootstrap-kubeconfig`.
+ 2. In the installer, sign the `kubelet-client` for ten years.  This is the actual bootstrap credential.
+ 3. In the cluster-kube-apiserver-operator, union in the content from `configmap/kubelet-bootstrap-kubeconfig`.
+ 4. In the MCO, if the `configmap/kubelet-bootstrap-kubeconfig` is deleted, use the "normal" service account token
+    logic that provides bootstrap authentication credentials for workers.
+ 5. The MCO would rollout the new machine configs and the kube-apiserver would effectively "revoke" the long lived client certificates.
+
+### Background about how kubelets start up
+
+ 1. If a current kubelet client certificate exists and is valid, use it and skip the rest.
+ 2. Use the bootstrap kubelet credentials to create a client CSR on the kube-apiserver.
+ 3. machine-approver (or cluster-admin) approvers the CSR created by the kubelet's bootstrap credentials.
+ 4. Kubelet uses bootstrap kubelet credentials to retrieve the signed certificate from the client CSR.
+ 
+If the bootstrap kubelet credentials are not valid (for instance, an expired certificate), then the kubelet is unable to create
+the CSR on the kube-apiserver and the kubelet remains unable to join the cluster.
+
+### Test Plan
+
+1. Unit tests within MCO validating the deletion of the configmaps and the creation of the bootstrap kubeconfig on masters
+2. e2e test within MCO testing the functionality within the cluster
+    1. Shall delete the configmap/kubelet-bootstrap-kubeconfig configmap and the cluster will roll out a managed kubeconfig to the masters with the token bootstrap credentials
+
+Wolfgang is willing to run manual testing of the shutdown/restore process.
+
+### Upgrade / Downgrade Strategy
+
+Because the `configmap/kubelet-bootstrap-kubeconfig` will not exist on upgrade,
+the machine config operator will update the machine config for masters to use a service account token instead.
+This will produce a second rollout of masters, one time.
+
+On downgrade, the machine config would revert back, but this is also ok because the masters would ripple out
+before the kubelet client certificates are lost.
+
+### Version Skew Strategy
+
+How will the component handle version skew with other components?
+What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- During an upgrade, we will always have skew among components, how will this impact your work?
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI, CRI
+  or CNI may require updating that component before the kubelet.
+
+## Drawbacks
+
+
+## Alternatives
+
+1. Use a service-account token from the beginning.
```

Comment:
we are in position to actually allow this ordering to happen with cluster etcd operator.. see https://github.com/openshift/machine-config-operator/issues/1427

So i think this enhancement should make stronger case for why not wait for this new order.

Reply:
> we are in position to actually allow this ordering to happen with cluster etcd operator.. see [openshift/machine-config-operator#1427](https://github.com/openshift/machine-config-operator/issues/1427)
> 
> So i think this enhancement should make stronger case for why not wait for this new order.

The etcd team is actually facing ordering and speed problems, so they're asking me to find a way to avoid using service account tokens.  While consistency would be good, I'm convinced of the safety and revocation options using this approach.  If it works, I plan to take similar action to help etcd achieve their objectives.